Add To/FromHttpApiData and To/FromFormKey instances for Const

carymrobbins opened the pull request:
As the description says, adds instances for `Const`.

phadej commented on src/Web/Internal/FormUrlEncoded.hs:
```diff
@@ -11,6 +12,7 @@
 {-# LANGUAGE OverloadedStrings          #-}
 {-# LANGUAGE PolyKinds                  #-}
 {-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE StandaloneDeriving         #-}
```
Cannot use, see how `Tagged` instance is written.

carymrobbins replied:
Happy to do it that way, but I'm curious; why must `Tagged` be written this way?

```haskell
instance ToHttpApiData a => ToHttpApiData (Tagged b a) where
  toUrlPiece        = coerce (toUrlPiece :: a -> Text)
  toHeader          = coerce (toHeader :: a -> ByteString)
  toQueryParam      = coerce (toQueryParam :: a -> Text)
  toEncodedUrlPiece = coerce (toEncodedUrlPiece ::  a -> BS.Builder)
```

phadej replied:
because `DerivingStrategies` is GHC-8.2+ feature, and I don't find the convenience to justify dropping support for older GHCs. You'll see CI failing soon.

carymrobbins replied:
Updated with a new commit.

phadej commented:
Enabling `PolyKinds` also made `Fixed` and `Tagged` instances polykinded. I have to think whether that's a breaking change.

Also with `text-show` flag the GHC-8.10 build is broken. (see CI).

carymrobbins commented:
> Enabling `PolyKinds` also made `Fixed` and `Tagged` instances polykinded. That is ok, and should been so from beginning. Yet, I have to think whether that's a breaking change.

It likely _would_ be a breaking change I guess since someone might have created an instance for another kind other than `*`.

> Also with `text-show` flag the GHC-8.10 build is broken. (see CI).

Will look into it.

phadej commented on src/Web/Internal/HttpApiData.hs:
```diff
@@ -5,7 +5,10 @@
 {-# LANGUAGE DeriveFunctor        #-}
 {-# LANGUAGE DeriveTraversable    #-}
 {-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
this is no longer necessary.

carymrobbins replied:
Oh nice catch

carymrobbins replied:
Updated

phadej commented:
> Also with text-show flag the GHC-8.10 build is broken. (see CI).

I think the problem is that `text-show` ought to have polykinded instance, and things would work.

carymrobbins commented:
> > Also with text-show flag the GHC-8.10 build is broken. (see CI).
> 
> I think the problem is that `text-show` ought to have polykinded instance, and things would work.

I could specialize it to use kind `*`. This could also be done for `Tagged` to avoid a breaking change. What do you think?

phadej commented:
maintainer of `text-show` would welcome a patch making `.Fixed` module `PolyKinds`-enabled.

phadej commented:
@fizruk do you have anything to say whether it would be simpler to use `text-show` always?

I don't see any downsides, and there would be another reason to make a breaking change major bump

phadej commented:
@carymrobbins to get this out the door, adding explicit kind signatures to Tagged and Fixed would be good options.

I use following trick:

```haskell
#if MIN_VERSION_base(4,9,0)
import Data.Kind (Type)
#else
#define Type *
```

and use `Type` in kind signatures. IIRC GHC-9.0 will start to have more warnings if you still use `*` in kinds.

carymrobbins commented:
> @carymrobbins to get this out the door, adding explicit kind signatures to Tagged and Fixed would be good options.
> 
> I use following trick:
> 
> ```haskell
> #if MIN_VERSION_base(4,9,0)
> import Data.Kind (Type)
> #else
> #define Type *
> #endif
> ```
> 
> and use `Type` in kind signatures. IIRC GHC-9.0 will start to have more warnings if you still use `*` in kinds.

Clever! Will do.

phadej commented:
last commit is wrong. Rather make non-polykinded instance for older GHCs.

carymrobbins commented:
> last commit is wrong. Rather make non-polykinded instance for older GHCs.

Which GHC min version should I use for the polykinded instances?

phadej commented:
> Which GHC min version should I use for the polykinded instances?

4.9

Also import it from `Control.Applicative` where it leaved before that.

carymrobbins commented:
> last commit is wrong. Rather make non-polykinded instance for older GHCs.

To clarify, do we want to make `Tagged` polykinded for `>= 4.9`? Or leave it specialized as `Type` regardless of the base version?

Also, it seems `Fixed` doesn't support polykinds, possibly due to the `HasResolution` constraint?

```
src/Web/Internal/HttpApiData.hs:468:55: error:
    • Expected kind ‘k’, but ‘a’ has kind ‘*’
    • In the first argument of ‘F.Fixed’, namely ‘(a :: k)’
      In the first argument of ‘ToHttpApiData’, namely
        ‘(F.Fixed (a :: k))’
      In the instance declaration for ‘ToHttpApiData (F.Fixed (a :: k))’
    |
468 | instance F.HasResolution a => ToHttpApiData (F.Fixed (a :: k)) where toUrlPiece = showt; toEncodedUrlPiece = unsafeToEncodedUrlPiece
    |                                                       ^
```

carymrobbins commented:
Finally got CI passing; required a CPP guard around `Identity` since that came around in base 4.8.

phadej commented:
`Data.Functor.Identity` was migrated from `transformers`. And we depend on it, so no need to have guards.

carymrobbins commented:
@phadej I think the PR should be all good now; can I get another review?

carymrobbins commented:
@phadej @fizruk Checking back in, anything stopping this from being merged?

fizruk commented:
@carymrobbins Looks great to me, except out-of-date branch! Let's wait one day to see if @phadej has anything to say?

Regarding always using `text-show` — at the time this package was created @gregwebs and @snoyberg advised (in #6) to make this dependency optional because it was considered a heavy dependency. I am not sure if same argument applies today, and although personally I have no reservations, this change probably deserves a separate discussion (in the mailing list?), giving users an opportunity to weigh in on whether they are all okay with `text-show` or not.

phadej commented:
Merged as https://github.com/fizruk/http-api-data/pull/103
